File: 2025-11/2025-11-18/100_characters.py
def one_hundred(chars):
    quotient, remainder = divmod(100, len(chars))

    return chars * quotient + chars[:remainder]

print(one_hundred("One hundred "))
print(one_hundred("freeCodeCamp "))
print(one_hundred("daily challenges "))
print(one_hundred("!"))

# --- TEST SUITE ---

tests_text = '''
1. one_hundred("One hundred ") should return "One hundred One hundred One hundred One hundred One hundred One hundred One hundred One hundred One ".
2. one_hundred("freeCodeCamp ") should return "freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeC".
3. one_hundred("daily challenges ") should return "daily challenges daily challenges daily challenges daily challenges daily challenges daily challenge".
4. one_hundred("!") should return "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 11 November/11-18/100Characters.js
function oneHundred(chars) {
/* Solution 1 (while())
  let characters = chars;

  while (characters.length < 100)
    characters += chars;

  return characters.slice(0, 100);
*/

/* Solution 2 (recursion)
  if (chars.length >= 100)
    return chars.slice(0, 100);

  return oneHundred(chars + chars);
*/

/* Solution 3 (.repeat())
  return chars.repeat(Math.ceil(100 / chars.length)).slice(0, 100);
*/

// Solution 4 (.padEnd())
  return chars.padEnd(100, chars);
}

console.log(oneHundred("One hundred "));
console.log(oneHundred("freeCodeCamp "));
console.log(oneHundred("daily challenges "));
console.log(oneHundred("!"));

// --- TEST SUITE ---

const testsText = `
1. oneHundred("One hundred ") should return "One hundred One hundred One hundred One hundred One hundred One hundred One hundred One hundred One ".
2. oneHundred("freeCodeCamp ") should return "freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeCamp freeCodeC".
3. oneHundred("daily challenges ") should return "daily challenges daily challenges daily challenges daily challenges daily challenges daily challenge".
4. oneHundred("!") should return "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!".
`;

const regex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(regex).map(match => match.groups)];

function runTests(testData) {
  console.log("---------------------------");
  console.log("🧪 Starting Verification...");
  console.log("---------------------------");

  let allPassed = true;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    if (functionCallOutput === testOutput) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉 SUCCESS: All tests PASSED.");
  else
    console.log("⚠️ WARNING: Some tests FAILED.");
}

runTests(testData);